src/code/z_skin.c

```diff
@@ -152,74 +183,90 @@ void func_800A5F60(GraphicsContext* gfxCtx, PSkinAwb* skin, s32 limbIndex, Gfx*
             gDPPipeSync(POLY_OPA_DISP++);
         }
     }
+
     CLOSE_DISPS(gfxCtx, "../z_skin.c", 433);
 }
 
-void func_800A60D8(Actor* actor, GlobalContext* globalCtx, PSkinAwb* skin, SkinCallback callback, SkinCallback2 arg4,
-                   s32 arg5, s32 arg6, s32 arg7) {
+void Skin_DrawImpl(Actor* actor, GlobalContext* globalCtx, Skin* skin, SkinPostLimbDraw postLimbDraw,
+                   SkinOverrideLimbDraw overrideLimbDraw, s32 setTranslation, s32 arg6, s32 drawFlags) {
     s32 i;
-    s32 dataType;
+    s32 segmentType;
     SkinLimb** skeleton;
     GraphicsContext* gfxCtx = globalCtx->state.gfxCtx;
     s32 pad;
 
     OPEN_DISPS(gfxCtx, "../z_skin.c", 471);
-    if ((arg7 & 1) == 0) {
-        func_800A6AC4(skin, &D_801600C0[0], actor, arg5);
+
+    if (!(drawFlags & SKIN_DRAW_FLAG_CUSTOM_TRANSFORMS)) {
+        Skin_ApplyAnimTransformations(skin, gSkinLimbMatrices, actor, setTranslation);
     }
+
     skeleton = SEGMENTED_TO_VIRTUAL(skin->skeletonHeader->segment);
-    if ((arg7 & 2) == 0) {
+
+    if (!(drawFlags & SKIN_DRAW_FLAG_CUSTOM_MATRIX)) {
         Mtx* mtx;
 
         gSPMatrix(POLY_OPA_DISP++, &gMtxClear, G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
         mtx = SkinMatrix_MtxFToNewMtx(gfxCtx, &skin->mtx);
+
         if (mtx == NULL) {
             goto close_disps;
         }
+
         gSPMatrix(POLY_OPA_DISP++, mtx, G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
     }
+
     for (i = 0; i < skin->skeletonHeader->limbCount; i++) {
-        s32 phi_a0 = 1;
+        s32 shouldDraw = true;
 
-        if (arg4 != NULL) {
-            phi_a0 = arg4(actor, globalCtx, i, skin);
+        if (overrideLimbDraw != NULL) {
+            shouldDraw = overrideLimbDraw(actor, globalCtx, i, skin);
         }
-        dataType = ((SkinLimb*)SEGMENTED_TO_VIRTUAL(skeleton[i]))->unk_8;
-        if (dataType == 4 && phi_a0 == 1) {
-            func_800A5E28(gfxCtx, skin, i, arg6, arg7);
-        } else if (dataType == 0xB && phi_a0 == 1) {
-            func_800A5F60(gfxCtx, skin, i, NULL, arg7);
+
+        segmentType = ((SkinLimb*)SEGMENTED_TO_VIRTUAL(skeleton[i]))->segmentType;
+
+        if (segmentType == SKIN_LIMB_TYPE_ANIMATED && shouldDraw == true) {
+            Skin_DrawAnimatedLimb(gfxCtx, skin, i, arg6, drawFlags);
+        } else if (segmentType == SKIN_LIMB_TYPE_NORMAL && shouldDraw == true) {
+            Skin_DrawLimb(gfxCtx, skin, i, NULL, drawFlags);
         }
     }
-    if (callback != NULL) {
-        callback(actor, globalCtx, skin);
+
+    if (postLimbDraw != NULL) {
+        postLimbDraw(actor, globalCtx, skin);
```

Comment:
oh yeah I didnt even notice that. will look at some usages of it and see if that name or another makes sense

Reply:
best part about this is they are pretty much universally used for logic updates after the horse is done drawing, but while the horses matrices are still in the matrix list.

Its primary job is updating colliders to match limb positions after they are drawn. and doing this logic in draw is exactly what leads to this glitch lol https://www.youtube.com/watch?v=VAqyqfgxnSk&t=852s

so, Im not sure if we want to call it postDraw after what it would be best to use it for, or name it after how it really gets used which is later logic updates

Reply:
postDraw sounds appropriate, postLimbDraw for skelanime often also updates colliders

the real issue with the callback is subscribing to the collision pools, not so much updating colliders
I wouldn't take the bug in consideration for naming it, imo it's misusage, and not intended usage for the callback

Reply:
I think it depends.
Currently we are considering this "logic update in draw" behaviour a bug (and it is labelled like that in every horse actor that does this). I think to be consistent with this being a bug then we should consider this callback to be a postDraw

Reply:
Btw, this behaviour of subscribing collisions during draw still exists in MM:
https://github.com/zeldaret/mm/pull/413/files#diff-1dac57939127037837e51b46d879fe422dfed7fb223d76d7b76fb4c6c4c12eb5R4694

Reply:
yeah his point is that the collider subscription is the specific bug, which is true and what the bug tags single out. updating the positions and stuff of colliders is fine, and normal actors do it as well. (i wouldnt say its common, but it happens. mostly with jntspheres)

but anyway yeah its fine, postDraw seems ok